SingularityService/src/main/java/com/hubspot/singularity/scheduler/SingularityUsagePoller.java

```diff
@@ -111,68 +114,146 @@ public void runActionOnPoll() {
-    for (SingularitySlaveUsage overloadedSlave : overLoadedSlavesByUsage) {
+
+    for (SingularitySlaveUsage overloadedSlave : overloadedSlavesByOverusage) {
       if (currentShuffleCleanupsTotal >= configuration.getMaxTasksToShuffleTotal()) {
         LOG.debug("Not shuffling any more tasks (totalShuffleCleanups: {})", currentShuffleCleanupsTotal);
         break;
       }
       int shuffledTasksOnSlave = 0;
-      List<TaskIdWithUsage> possibleTasksToShuffle = overLoadedHosts.get(overloadedSlave);
-      possibleTasksToShuffle.sort((u1, u2) ->
-          Double.compare(
-              u2.getUsage().getCpusUsed() / u2.getRequestedResources().getCpus(),
-              u1.getUsage().getCpusUsed() / u1.getRequestedResources().getCpus()
-          ));
 
-      double systemLoad = getSystemLoadForShuffle(overloadedSlave);
-      double cpuOverage = systemLoad - overloadedSlave.getSystemCpusTotal();
+      double currentCpuLoad = getSystemLoadForShuffle(overloadedSlave);
+      double currentMemUsageBytes = overloadedSlave.getMemoryBytesUsed();
+
+      OverusedResource mostOverusedResource = getMostOverusedResource(overloadedSlave, currentCpuLoad, currentMemUsageBytes);
+
+      List<TaskIdWithUsage> possibleTasksToShuffle;
+      boolean shufflingForCpu;
+      if (mostOverusedResource.resourceType == Type.CPU) {
+        shufflingForCpu = true;
+        possibleTasksToShuffle = overLoadedHosts.get(overloadedSlave);
+        possibleTasksToShuffle.sort((u1, u2) ->
+            Double.compare(
+                u2.getUsage().getCpusUsed() / u2.getRequestedResources().getCpus(),
+                u1.getUsage().getCpusUsed() / u1.getRequestedResources().getCpus()
+            ));
+      } else {
+        shufflingForCpu = false;
+        possibleTasksToShuffle = overLoadedHosts.get(overloadedSlave);
+        possibleTasksToShuffle.sort((u1, u2) ->
+            Double.compare(
+                u2.getUsage().getMemoryTotalBytes() / u2.getRequestedResources().getMemoryMb(),
+                u1.getUsage().getMemoryTotalBytes() / u1.getRequestedResources().getMemoryMb()
+            ));
+      }
 
       for (TaskIdWithUsage taskIdWithUsage : possibleTasksToShuffle) {
         if (requestsWithShuffledTasks.contains(taskIdWithUsage.getTaskId().getRequestId())) {
           LOG.debug("Request {} already has a shuffling task, skipping", taskIdWithUsage.getTaskId().getRequestId());
           continue;
         }
-        if (cpuOverage <= 0 || shuffledTasksOnSlave > configuration.getMaxTasksToShufflePerHost() || currentShuffleCleanupsTotal >= configuration.getMaxTasksToShuffleTotal()) {
-          LOG.debug("Not shuffling any more tasks (overage: {}, shuffledOnHost: {}, totalShuffleCleanups: {})", cpuOverage, shuffledTasksOnSlave, currentShuffleCleanupsTotal);
+        if ((mostOverusedResource.overusage <= 0) || shuffledTasksOnSlave > configuration.getMaxTasksToShufflePerHost() || currentShuffleCleanupsTotal >= configuration.getMaxTasksToShuffleTotal()) {
```

Comment:
`cpuOverage <= 0` worked as a condition for 'should this put us back in green?' because we updated it at the end of the loop. I don't currently see `mostOverusedResource.overusage` getting updated anywhere